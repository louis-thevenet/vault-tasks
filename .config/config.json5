{
    "tasks_config": {
        "use_american_format": true,
        "indent_length": 2,
        "parse_dot_files": false,
        "vault_path": "./test-vault",
        "ignored": [],
    },
    "keybindings": {
        "Home": {
            "<q>": "Quit",
            "<Ctrl-d>": "Quit",
            "<Ctrl-c>": "Quit",
            "<Ctrl-z>": "Suspend",
            "<Ctrl-Right>":"TabRight",
            "<Ctrl-l>":"TabRight",
            "<Ctrl-Left>":"TabLeft",
            "<Ctrl-h>":"TabLeft",
        },
        "Tags": {
            "<q>": "Quit",
            "<Ctrl-d>": "Quit",
            "<Ctrl-c>": "Quit",
            "<Ctrl-z>": "Suspend",

            "<Ctrl-Right>":"TabRight",
            "<Ctrl-l>":"TabRight",
            "<Ctrl-Left>":"TabLeft",
            "<Ctrl-h>":"TabLeft",
        },
        "Explorer": {
            "<q>": "Quit",
            "<Ctrl-d>": "Quit",
            "<Ctrl-c>": "Quit",
            "<Ctrl-z>": "Suspend",

            "<Ctrl-Right>":"TabRight",
            "<Ctrl-l>":"TabRight",
            "<Ctrl-Left>":"TabLeft",
            "<Ctrl-h>":"TabLeft",

            "<j>": "Down",
            "<Down>": "Down",
            "<Tab>": "Down",
            "<k>": "Up",
            "<Up>": "Up",
            "<BackTab>": "Up",
            "<h>": "Left",
            "<Left>": "Left",
            "<l>": "Right",
            "<Right>": "Right",
            "<Enter>": "Enter",
            "<Backspace>": "Cancel",
        },
    },
}
