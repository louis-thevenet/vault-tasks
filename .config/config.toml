[keybindings.Calendar]
# App
"<q>" = "Quit"
"<Ctrl-c>" = "Quit"
"<Ctrl-z>" = "Suspend"
"<?>" = "Help"
# Tabs
"<Shift-Right>" = "TabRight"
"<Shift-l>" = "TabRight"
"<Shift-Left>" = "TabLeft"
"<shift-h>" = "TabLeft"
# Scrolling
"<Ctrl-u>" = "ViewUp"
"<Ctrl-k>" = "ViewUp"
"<Ctrl-Up>" = "ViewUp"
"<PageUp>" = "ViewPageUp"
"<Ctrl-j>" = "ViewDown"
"<Ctrl-d>" = "ViewDown"
"<Ctrl-Down>" = "ViewDown"
"<PageDown>" = "ViewPageDown"
# Navigation
"<r>" = "ReloadVault"
"<t>" = "GotoToday"
"<j>" = "Down"
"<Down>" = "Down"
"<k>" = "Up"
"<Up>" = "Up"
"<h>" = "Left"
"<Left>" = "Left"
"<l>" = "Right"
"<Right>" = "Right"
"<Shift-j>" = "NextMonth"
"<Shift-Down>" = "NextMonth"
"<Shift-k>" = "PreviousMonth"
"<Shift-Up>" = "PreviousMonth"
"<n>" = "NextYear"
"<Shift-n>" = "PreviousYear"

[keybindings.Explorer]
# App
"<q>" = "Quit"
"<Ctrl-c>" = "Quit"
"<Ctrl-z>" = "Suspend"
"<?>" = "Help"
# Tabs
"<Shift-Right>" = "TabRight"
"<Shift-l>" = "TabRight"
"<Shift-Left>" = "TabLeft"
"<shift-h>" = "TabLeft"
# Navigation
"<j>" = "Down"
"<Down>" = "Down"
"<Tab>" = "Down"
"<k>" = "Up"
"<Up>" = "Up"
"<BackTab>" = "Up"
"<h>" = "Left"
"<Left>" = "Left"
"<l>" = "Right"
"<Right>" = "Right"
"<Enter>" = "Enter"
"<Backspace>" = "Cancel"
"<s>" = "Search"
"<Esc>" = "Escape"
"<o>" = "Open"
"<e>" = "Edit"
"<t>" = "MarkToDo"
"<d>" = "MarkDone"
"<c>" = "MarkCancel"
"<i>" = "MarkIncomplete"
"<r>" = "ReloadVault"
"<+>" = "IncreaseCompletion"
"<->" = "DecreaseCompletion"
# Scrolling
"<Ctrl-u>" = "ViewUp"
"<Ctrl-k>" = "ViewUp"
"<Ctrl-Up>" = "ViewUp"
"<PageUp>" = "ViewPageUp"
"<Ctrl-j>" = "ViewDown"
"<Ctrl-d>" = "ViewDown"
"<Ctrl-Down>" = "ViewDown"
"<PageDown>" = "ViewPageDown"
"<Ctrl-l>" = "ViewRight"
"<Ctrl-Right>" = "ViewRight"
"<Ctrl-h>" = "ViewLeft"
"<Ctrl-Left>" = "ViewLeft"

[keybindings.Filter]
# App
"<q>" = "Quit"
"<Ctrl-c>" = "Quit"
"<Ctrl-z>" = "Suspend"
"<?>" = "Help"
# Tabs
"<Shift-Right>" = "TabRight"
"<Shift-l>" = "TabRight"
"<Shift-Left>" = "TabLeft"
"<shift-h>" = "TabLeft"
# Navigation
"<Enter>" = "Enter"
"<s>" = "Search"
"<Shift-s>" = "SwitchSortingMode"
"<Esc>" = "Escape"
"<r>" = "ReloadVault"
# Scrolling
"<Ctrl-u>" = "ViewUp"
"<Ctrl-k>" = "ViewUp"
"<Ctrl-Up>" = "ViewUp"
"<PageUp>" = "ViewPageUp"
"<Ctrl-d>" = "ViewDown"
"<Ctrl-j>" = "ViewDown"
"<Ctrl-Down>" = "ViewDown"
"<PageDown>" = "ViewPageDown"
"<Ctrl-l>" = "ViewRight"
"<Ctrl-Right>" = "ViewRight"
"<Ctrl-h>" = "ViewLeft"
"<Ctrl-Left>" = "ViewLeft"

[keybindings.Home]
# App
"<q>" = "Quit"
"<Ctrl-d>" = "Quit"
"<Ctrl-c>" = "Quit"
"<Ctrl-z>" = "Suspend"
# Tabs
"<Shift-Right>" = "TabRight"
"<Shift-l>" = "TabRight"
"<Shift-Left>" = "TabLeft"
"<Shift-h>" = "TabLeft"

[keybindings.TimeManagement]
# App
"<q>" = "Quit"
"<Ctrl-c>" = "Quit"
"<Ctrl-z>" = "Suspend"
"<?>" = "Help"
# Tabs
"<Shift-Right>" = "TabRight"
"<Shift-l>" = "TabRight"
"<Shift-Left>" = "TabLeft"
"<shift-h>" = "TabLeft"
# Navigation
"<e>" = "Edit"
"<Esc>" = "Escape"
"<Enter>" = "Enter"
"<Space>" = "NextSegment"
"<p>" = "Pause"
"<Tab>" = "NextMethod"
"<BackTab>" = "PreviousMethod"
"<j>" = "Down"
"<Down>" = "Down"
"<k>" = "Up"
"<Up>" = "Up"
"<h>" = "Left"
"<Left>" = "Left"
"<l>" = "Right"
"<Right>" = "Right"

[styles.Explorer]
preview_headers = "bold rgb 255 153 000"

[styles.Home]
highlighted_style = "dark grey on rgb 255 153 000"
highlighted_bar_style = "rgb 255 153 000"

[tasks_config]
use_american_format = true
show_relative_due_dates = true
indent_length = 2
parse_dot_files = false
file_tags_propagation = true
ignored = []
# vault_path= "./test-vault" # default vault path when none is provided
explorer_default_search_string = "- [ ] "
filter_default_search_string = ""
completion_bar_length = 5
tracker_extra_blanks = 3

# Nice if your terminal font doesn't have emojis
# pretty_symbols.task_done="[x]"
# pretty_symbols.task_todo="[ ]"
# pretty_symbols.task_incomplete="[/]"
# pretty_symbols.task_canceled="[-]"
# pretty_symbols.due_date="@"
# pretty_symbols.priority="!"
# pretty_symbols.today_tag="+"
# pretty_symbols.progress_bar_true="="
# pretty_symbols.progress_bar_false=" "

pretty_symbols.task_done = "‚úÖ"
pretty_symbols.task_todo = "‚ùå"
pretty_symbols.task_incomplete = "‚è≥"
pretty_symbols.task_canceled = "üö´"
pretty_symbols.due_date = "üìÖ"
pretty_symbols.priority = "‚ùó"
pretty_symbols.today_tag = "‚òÄÔ∏è"
pretty_symbols.progress_bar_true = "üü©"
pretty_symbols.progress_bar_false = "‚¨úÔ∏è"

task_state_markers.todo = ' '
task_state_markers.done = 'x'
task_state_markers.incomplete = '/'
task_state_markers.canceled = '-'

[[time_management_methods_settings.FlowTime]]
name = "Break Factor"
hint = "Break time is (focus time) / (break factor)"

[time_management_methods_settings.FlowTime.value]
Int = 5

[[time_management_methods_settings.FlowTime]]
name = "Auto Skip"
hint = "Whether to wait for user input to enter new segments or not"

[time_management_methods_settings.FlowTime.value]
Bool = false

[[time_management_methods_settings.Pomodoro]]
name = "Focus Time"
hint = ""

[time_management_methods_settings.Pomodoro.value.Duration]
secs = 1500
nanos = 0

[[time_management_methods_settings.Pomodoro]]
name = "Short Break Time"
hint = ""

[time_management_methods_settings.Pomodoro.value.Duration]
secs = 300
nanos = 0

[[time_management_methods_settings.Pomodoro]]
name = "Auto Skip"
hint = "Whether to wait for user input to enter new segments or not"

[time_management_methods_settings.Pomodoro.value]
Bool = false

[[time_management_methods_settings.Pomodoro]]
name = "Long Break Time"
hint = ""

[time_management_methods_settings.Pomodoro.value.Duration]
secs = 900
nanos = 0

[[time_management_methods_settings.Pomodoro]]
name = "Long Break Interval"
hint = "Short breaks before a long break"

[time_management_methods_settings.Pomodoro.value]
Int = 4
