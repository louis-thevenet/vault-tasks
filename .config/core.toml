[core]
parse_dot_files = false
file_tags_propagation = true
ignored = []
# default vault paths when none is provided
vault_paths = [
  # "./test-vault",
]
use_american_format = true
indent_length = 2
tracker_extra_blanks = 3

[display]
show_relative_due_dates = true

[pretty_symbols]
task_done = "✅"
task_todo = "❌"
task_incomplete = "⏳"
task_canceled = "🚫"
due_date = "📅"
priority = "❗"
today_tag = "☀️"
progress_bar_true = "🟩"
progress_bar_false = "⬜️"

[task_state_markers]
todo = ' '
done = 'x'
incomplete = '/'
canceled = '-'

# Nice if your terminal font doesn't have emojis
# task_done="[x]"
# task_todo="[ ]"
# task_incomplete="[/]"
# task_canceled="[-]"
# due_date="@"
# priority="!"
# today_tag="+"
# progress_bar_true="="
# progress_bar_false=" "
