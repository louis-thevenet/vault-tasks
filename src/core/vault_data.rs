use std::fmt::Display;

use super::task::Task;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VaultData {
    /// Name, Content
    Directory(String, Vec<VaultData>),
    /// Name, Content
    Header(usize, String, Vec<VaultData>),
    /// Task, Subtasks
    Task(Task),
}

impl Display for VaultData {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fn write_indent(indent_length: usize, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            (1..=indent_length).try_for_each(|_| write!(f, "\t"))?;
            Ok(())
        }
        fn write_underline_with_indent(
            text: &str,
            indent_length: usize,
            f: &mut std::fmt::Formatter,
        ) -> std::fmt::Result {
            write_indent(indent_length, f)?;
            writeln!(f, "{text}")?;
            write_indent(indent_length, f)?;
            for _i in 0..(text.len()) {
                write!(f, "â€¾")?;
            }
            writeln!(f)?;
            Ok(())
        }
        fn fmt_aux(
            file_entry: &VaultData,
            f: &mut std::fmt::Formatter,
            depth: usize,
        ) -> std::fmt::Result {
            match file_entry {
                VaultData::Header(_, header, entries) => {
                    write_underline_with_indent(&header.to_string(), depth, f)?;
                    for entry in entries {
                        fmt_aux(entry, f, depth + 1)?;
                    }
                }
                VaultData::Directory(name, entries) => {
                    write_underline_with_indent(&name.to_string(), depth, f)?;
                    for entry in entries {
                        fmt_aux(entry, f, depth + 1)?;
                    }
                }
                VaultData::Task(task) => {
                    for line in task.to_string().split('\n') {
                        write_indent(depth, f)?;
                        writeln!(f, "{line}")?;
                    }

                    for subtask in &task.subtasks {
                        for line in VaultData::Task(subtask.clone()).to_string().split('\n') {
                            write_indent(depth + 1, f)?;
                            writeln!(f, "{line}")?;
                        }
                    }
                }
            }
            Ok(())
        }
        fmt_aux(self, f, 0)
    }
}
