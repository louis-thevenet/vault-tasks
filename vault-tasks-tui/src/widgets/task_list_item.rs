use std::rc::Rc;

use ratatui::{
    prelude::*,
    widgets::{Block, Borders, Cell, Paragraph, Row, Table},
};
use ratskin::RatSkin;
use tracing::error;
use vault_tasks_core::{config::TasksConfig, task::Task, tracker::Tracker, vault_data::VaultData};

const HEADER_INDENT_RATIO: u16 = 3;

#[derive(Clone)]
pub struct TaskListItem {
    item: VaultData,
    pub height: u16,
    // TODO: here we use stuff that should be in TUI and not core
    config: TasksConfig,
    max_width: u16,
    display_filename: bool,
    header_style: Style,
}

impl TaskListItem {
    pub fn header_style(mut self, style: Style) -> Self {
        self.header_style = style;
        self
    }
    pub fn new(
        item: VaultData,
        config: TasksConfig,
        max_width: u16,
        display_filename: bool,
    ) -> Self {
        let height = Self::compute_height(&item, max_width);
        Self {
            item,
            height,
            config,
            max_width,
            display_filename,
            header_style: Style::default(),
        }
    }
    #[allow(clippy::too_many_lines)]
    fn task_to_paragraph(&self, area: Rect, task: &Task) -> (Rc<[Rect]>, Paragraph<'_>) {
        let mut lines = vec![];
        let mut data_line = vec![];

        let rat_skin = RatSkin::default();

        let state = task.state.display(self.config.pretty_symbols.clone());
        let title = state.clone() + " " + &task.name;
        let title_parsed = rat_skin.parse(RatSkin::parse_text(&title), self.max_width);
        let binding = Line::raw(state);
        let title = match title_parsed.first() {
            Some(t) => {
                lines.append(&mut title_parsed[1..].to_vec());
                t
            }
            None => &binding,
        };

        let surrounding_block =
            Block::default()
                .borders(Borders::ALL)
                .title_bottom(if self.display_filename {
                    Line::from(task.filename.clone()).right_aligned()
                } else {
                    Line::from("")
                });

        if task.is_today {
            data_line.push(Span::raw(format!(
                "{} ",
                self.config.pretty_symbols.today_tag
            )));
        }

        if let Some(due_date) = &task.due_date {
            data_line.push(Span::from(format!(
                "{} ",
                due_date.to_display_format(
                    &self.config.pretty_symbols.due_date,
                    !self.config.use_american_format,
                )
            )));
            if self.config.show_relative_due_dates {
                let due_date_relative = due_date.get_relative_str();
                data_line.push(Span::styled(
                    format!("({due_date_relative}) "),
                    Style::new().dim(),
                ));
            }
        }
        if let Some(bar) = task.get_completion_bar(
            self.config.completion_bar_length,
            &(
                self.config.pretty_symbols.progress_bar_false.clone(),
                self.config.pretty_symbols.progress_bar_true.clone(),
            ),
        ) {
            data_line.push(Span::raw(bar));
        }
        if task.priority > 0 {
            data_line.push(Span::raw(format!(
                "{}{} ",
                self.config.pretty_symbols.priority, task.priority
            )));
        }
        if !data_line.is_empty() {
            lines.push(Line::from(data_line));
        }
        let mut tag_line = String::new();
        if task.tags.is_some() {
            tag_line.push_str(
                &task
                    .tags
                    .clone()
                    .unwrap()
                    .iter()
                    .map(|t| format!("#{t}"))
                    .collect::<Vec<String>>()
                    .join(" "),
            );
        }
        if !tag_line.is_empty() {
            lines.push(Line::from(Span::styled(tag_line, Color::DarkGray)));
        }
        if let Some(description) = task.description.clone() {
            let text = rat_skin.parse(RatSkin::parse_text(&description), self.max_width);
            lines = [lines, text].concat();
        }
        let mut constraints = vec![Constraint::Length((lines.len()).try_into().unwrap())];

        for st in &task.subtasks {
            constraints.push(Constraint::Length(Self::compute_height(
                &VaultData::Task(st.clone()),
                self.max_width - 2, // -2 for borders
            )));
        }

        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints(constraints)
            .split(surrounding_block.inner(area));

        (
            layout,
            if lines.is_empty() && task.subtasks.is_empty() {
                Paragraph::new(title.clone()).block(surrounding_block)
            } else {
                Paragraph::new(Text::from(lines)).block(surrounding_block.title_top(title.clone()))
            },
        )
    }
    fn tracker_to_table(&self, tracker: &Tracker) -> Table<'_> {
        let header = [
            vec!["Dates".to_owned()],
            tracker.categories.iter().map(|c| c.name.clone()).collect(),
        ]
        .concat()
        .into_iter()
        .map(Cell::from)
        .collect::<Row>()
        .style(self.header_style)
        .height(1);

        let mut date = tracker.start_date.clone();
        let rat_skin = RatSkin::default();
        let mut rows = (0..tracker.length)
            .map(|n| {
                let res = Row::new(
                    [
                        vec![Cell::from(
                            Span::raw(
                                date.to_string_format(!self.config.use_american_format)
                                    .to_string(),
                            ) + if self.config.show_relative_due_dates {
                                Span::raw(format!(" ({})", date.get_relative_str())).dim()
                            } else {
                                Span::raw("")
                            },
                        )],
                        tracker
                            .categories
                            .iter()
                            .map(|c| {
                                Cell::from(Text::from(
                                    rat_skin.parse(
                                        RatSkin::parse_text(
                                            &c.entries
                                                .get(n)
                                                .unwrap()
                                                .pretty_fmt(&self.config.pretty_symbols),
                                        ),
                                        self.max_width,
                                    ),
                                ))
                            })
                            .collect(),
                    ]
                    .concat(),
                );
                date = tracker.frequency.next_date(&date);
                res
            })
            .collect::<Vec<Row>>();
        if self.config.invert_tracker_entries {
            rows.reverse();
        }
        let mut date = tracker.start_date.clone();
        let widths = [
            vec![
                (0..tracker.length)
                    .map(|_n| {
                        let res = if self.config.show_relative_due_dates {
                            format!(
                                "{} ({})",
                                date.to_string_format(!self.config.use_american_format,),
                                date.get_relative_str()
                            )
                        } else {
                            date.to_string_format(!self.config.use_american_format)
                                .to_string()
                        }
                        .len() as u16;
                        date = tracker.frequency.next_date(&date);
                        res
                    })
                    .max()
                    .unwrap_or_default(),
            ],
            tracker
                .categories
                .iter()
                .map(|cat| {
                    (cat.entries
                        .iter()
                        .map(|ent| ent.to_string().len())
                        .max()
                        .unwrap_or_default())
                    .max(cat.name.len()) as u16
                })
                .collect::<Vec<u16>>(),
        ]
        .concat();
        Table::new(rows, widths)
            .header(header)
            .column_spacing(2)
            .block(Block::bordered().title(tracker.name.clone()))
    }
    fn compute_height(item: &VaultData, max_width: u16) -> u16 {
        let rat_skin = RatSkin::default();
        match &item {
            VaultData::Directory(_, _) => 1,
            VaultData::Header(_, _, children) => {
                children
                    .iter()
                    .map(|c| Self::compute_height(c, max_width * (100 - HEADER_INDENT_RATIO) / 100))
                    .sum::<u16>()
                    + 1 // name in block (border only on top)
            }
            VaultData::Task(task) => {
                let mut count: u16 = 2; // block
                if 2 + task.name.len() >= max_width as usize {
                    count += (2 + task.name.len() as u16) / max_width;
                }
                if let Some(d) = &task.description {
                    let parsed_desc = rat_skin.parse(RatSkin::parse_text(d), max_width);
                    count += u16::try_from(parsed_desc.len()).unwrap_or_else(|e| {
                        error!("Could not convert description length to u16 :{e}");
                        0
                    });
                }
                if task.due_date.is_some()
                    || task.priority > 0
                    || task.is_today
                    || task.completion.is_some()
                {
                    count += 1;
                }
                if task.tags.is_some() {
                    count += 1;
                }
                for sb in &task.subtasks {
                    count += Self::compute_height(&VaultData::Task(sb.clone()), max_width - 2);
                }
                count.max(3) // If count == 2 then task name will go directly inside a block
                // Else task name will be the block's title and content will go inside
            }
            VaultData::Tracker(tracker) => {
                2 // block
                    + 1 // header
                    + tracker
                    .categories
                    .first()
                    .map_or(0, |c| c.entries.len() as u16) // number of entries
            }
        }
    }
}
impl Widget for TaskListItem {
    fn render(self, area: Rect, buf: &mut Buffer)
    where
        Self: Sized,
    {
        let rat_skin = RatSkin::default();
        match &self.item {
            VaultData::Directory(name, _) => error!("TaskList widget received a directory: {name}"),
            VaultData::Header(_level, name, children) => {
                let surrounding_block = Block::default().borders(Borders::TOP).title(
                    rat_skin
                        .parse(RatSkin::parse_text(name), area.width)
                        .first()
                        .unwrap()
                        .clone()
                        .style(self.header_style),
                );

                let indent = Layout::new(
                    Direction::Horizontal,
                    vec![
                        Constraint::Percentage(HEADER_INDENT_RATIO),
                        Constraint::Percentage(100 - HEADER_INDENT_RATIO),
                    ],
                )
                .split(area);

                let mut constraints = vec![];
                for child in children {
                    constraints.push(Constraint::Length(Self::compute_height(
                        child,
                        self.max_width - indent[0].width,
                    )));
                }
                let layout = Layout::default()
                    .direction(Direction::Vertical)
                    .constraints(constraints)
                    .split(surrounding_block.inner(indent[1]));
                surrounding_block.render(area, buf);

                for (i, child) in children.iter().enumerate() {
                    let sb_widget = Self::new(
                        child.clone(),
                        self.config.clone(),
                        self.max_width - indent[0].width,
                        self.display_filename,
                    )
                    .header_style(self.header_style);
                    sb_widget.render(layout[i], buf);
                }
            }
            VaultData::Task(task) => {
                let (layout, par) = self.task_to_paragraph(area, task);
                par.render(area, buf);

                for (i, sb) in task.subtasks.iter().enumerate() {
                    let sb_widget = Self::new(
                        VaultData::Task(sb.clone()),
                        self.config.clone(),
                        self.max_width - 2, // surrounding block
                        false,
                    )
                    .header_style(self.header_style);

                    sb_widget.render(layout[i + 1], buf);
                }
            }
            VaultData::Tracker(tracker) => {
                Widget::render(self.tracker_to_table(tracker), area, buf);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use chrono::NaiveDate;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};
    use vault_tasks_core::{
        date::Date,
        task::{State, Task},
        vault_data::VaultData,
    };

    use crate::{config::Config, widgets::task_list_item::TaskListItem};

    #[test]
    fn test_task_list_item() {
        let test_task = VaultData::Task(Task {
            name: "task with a very long title that should wrap to the next line".to_string(),
            state: State::Done,
            tags: Some(vec![String::from("tag"), String::from("tag2")]),
            priority: 5,
            completion: Some(60),
            due_date: Some(Date::DayTime(
                NaiveDate::from_ymd_opt(2016, 7, 8)
                    .unwrap()
                    .and_hms_opt(9, 10, 11)
                    .unwrap(),
            )),
            subtasks: vec![
                Task {
                    name: "subtask with another long title that should wrap around".to_string(),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                },
                Task {
                    name: "subtask test".to_string(),
                    tags: Some(vec![String::from("tag"), String::from("tag2")]),
                    ..Default::default()
                },
                Task {
                    name: "subtask test with a long title 123456789 1 2 3".to_string(),
                    priority: 5,
                    due_date: Some(Date::DayTime(
                        NaiveDate::from_ymd_opt(2016, 7, 8)
                            .unwrap()
                            .and_hms_opt(9, 10, 11)
                            .unwrap(),
                    )),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                },
            ],
            ..Default::default()
        });
        let mut config = Config::default();

        // We don't want tests to be time dependent
        config.tasks_config.show_relative_due_dates = false;

        let max_width = 50;
        let task_list_item =
            TaskListItem::new(test_task, config.tasks_config.clone(), max_width, false);
        let mut terminal = Terminal::new(TestBackend::new(max_width, 40)).unwrap();
        terminal
            .draw(|frame| {
                frame.render_widget(task_list_item, frame.area());
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
