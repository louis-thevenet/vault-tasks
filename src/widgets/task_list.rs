use crate::core::vault_data::VaultData;
use ratatui::prelude::*;
use tui_scrollview::{ScrollView, ScrollViewState};

use crate::config::Config;

use super::task_list_item::TaskListItem;

#[derive(Default, Clone)]
pub struct TaskList {
    content: Vec<TaskListItem>,
    constraints: Vec<Constraint>,
    height: u16,
}

impl TaskList {
    pub fn new(
        config: &Config,
        file_content: &[VaultData],
        max_width: u16,
        display_filename: bool,
    ) -> Self {
        let content = file_content
            .iter()
            .map(|fc| {
                TaskListItem::new(
                    fc.clone(),
                    !config.tasks_config.use_american_format,
                    config.tasks_config.pretty_symbols.clone(),
                    max_width,
                    display_filename,
                    config.tasks_config.show_relative_due_dates,
                    config.tasks_config.completion_bar_length,
                )
                .header_style(
                    *config
                        .styles
                        .get(&crate::app::Mode::Explorer)
                        .unwrap()
                        .get("preview_headers")
                        .unwrap(),
                )
            })
            .collect::<Vec<TaskListItem>>();
        let mut height = 0;
        let mut constraints = vec![];
        for item in &content {
            height += item.height;
            constraints.push(Constraint::Length(item.height));
        }
        Self {
            content,
            constraints,
            height,
        }
    }
    // pub fn height_of(&mut self, i: usize) -> u16 {
    //     (0..i).map(|i| self.content[i].height).sum()
    // }
}
impl StatefulWidget for TaskList {
    type State = ScrollViewState;
    fn render(
        self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &mut Self::State,
    ) where
        Self: Sized,
    {
        // If we need the vertical scrollbar
        // Then take into account that we need to draw it
        //
        // If we don't do this, the horizontal scrollbar
        // appears for only one character
        // It basically disables the horizontal scrollbar
        let width = if self.height > area.height {
            area.width - 1
        } else {
            area.width
        };

        let size = Size::new(width, self.height);
        let mut scroll_view = ScrollView::new(size);

        let layout = Layout::vertical(self.constraints).split(scroll_view.area());

        for (i, item) in self.content.into_iter().enumerate() {
            scroll_view.render_widget(item, layout[i]);
        }
        scroll_view.render(area, buf, state);
    }
}

#[cfg(test)]
mod tests {
    use crate::core::{
        date::DueDate,
        task::{State, Task},
        vault_data::VaultData,
    };
    use chrono::NaiveDate;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};
    use tui_scrollview::ScrollViewState;

    use crate::{config::Config, widgets::task_list::TaskList};

    #[test]
    fn test_task_list() {
        let test_vault = VaultData::Header(
            0,
            "Test".to_string(),
            vec![
                VaultData::Header(
                    1,
                    "1".to_string(),
                    vec![
                        VaultData::Task(Task {
                            name: "task 1".to_string(),
                            state: State::Done,
                            tags: Some(vec![String::from("tag"), String::from("tag2")]),
                            priority: 5,
                            due_date: DueDate::DayTime(
                                NaiveDate::from_ymd_opt(2016, 7, 8)
                                    .unwrap()
                                    .and_hms_opt(9, 10, 11)
                                    .unwrap(),
                            ),
                            subtasks: vec![
                                Task {
                                    name: "subtask test with desc".to_string(),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                },
                                Task {
                                    name: "subtask test with tags".to_string(),
                                    tags: Some(vec![String::from("tag"), String::from("tag2")]),
                                    ..Default::default()
                                },
                                Task {
                                    name: "subtask test".to_string(),
                                    ..Default::default()
                                },
                            ],
                            ..Default::default()
                        }),
                        VaultData::Header(
                            2,
                            "1.1".to_string(),
                            vec![VaultData::Header(
                                3,
                                "1.1.1".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "test 1.1.1".to_string(),
                                    description: Some("test\ndesc\nðŸ¥ƒ".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                    ],
                ),
                VaultData::Header(
                    1,
                    "2".to_string(),
                    vec![
                        VaultData::Header(3, "2.1".to_string(), vec![]),
                        VaultData::Header(
                            2,
                            "2.2".to_string(),
                            vec![VaultData::Task(Task {
                                name: "test 2.2".to_string(),
                                description: Some("test\ndesc".to_string()),
                                subtasks: vec![Task {
                                    name: "subtask 2.2".to_string(),

                                    due_date: DueDate::DayTime(
                                        NaiveDate::from_ymd_opt(2016, 7, 8)
                                            .unwrap()
                                            .and_hms_opt(9, 10, 11)
                                            .unwrap(),
                                    ),
                                    description: Some("test\ndesc".to_string()),
                                    tags: Some(vec![String::from("tag"), String::from("tag2")]),
                                    ..Default::default()
                                }],
                                ..Default::default()
                            })],
                        ),
                    ],
                ),
            ],
        );

        let mut config = Config::default();

        // We don't want tests to be time dependent
        config.tasks_config.show_relative_due_dates = false;

        let max_width = 40;
        let task_list = TaskList::new(&config, &[test_vault], max_width, true);
        let mut terminal = Terminal::new(TestBackend::new(max_width, 40)).unwrap();
        terminal
            .draw(|frame| {
                frame.render_stateful_widget(task_list, frame.area(), &mut ScrollViewState::new());
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
