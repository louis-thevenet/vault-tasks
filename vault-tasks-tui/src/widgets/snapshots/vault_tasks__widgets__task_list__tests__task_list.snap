---
source: src/widgets/task_list.rs
expression: terminal.backend()
---
"Test────────────────────────────────────"
" 1──────────────────────────────────────"
"  ┌✅ task 1───────────────────────────┐" Hidden by multi-width symbols: [(4, " ")]
"  │📅 2016-07-08 09:10:11 ❗5          │" Hidden by multi-width symbols: [(4, " "), (27, " ")]
"  │#tag #tag2                          │"
"  │┌❌ subtask test with desc─────────┐│" Hidden by multi-width symbols: [(5, " ")]
"  ││test                              ││"
"  ││desc                              ││"
"  │└──────────────────────────────────┘│"
"  │┌❌ subtask test with tags─────────┐│" Hidden by multi-width symbols: [(5, " ")]
"  ││#tag #tag2                        ││"
"  │└──────────────────────────────────┘│"
"  │┌──────────────────────────────────┐│"
"  ││❌ subtask test                   ││" Hidden by multi-width symbols: [(5, " ")]
"  │└──────────────────────────────────┘│"
"  └────────────────────────────────────┘"
"  1.1───────────────────────────────────"
"   1.1.1────────────────────────────────"
"    ┌❌ test 1.1.1─────────────────────┐" Hidden by multi-width symbols: [(6, " ")]
"    │test                              │"
"    │desc                              │"
"    │🥃                                │" Hidden by multi-width symbols: [(6, " ")]
"    └──────────────────────────────────┘"
" 2──────────────────────────────────────"
"  2.1───────────────────────────────────"
"  2.2───────────────────────────────────"
"   ┌❌ test 2.2────────────────────────┐" Hidden by multi-width symbols: [(5, " ")]
"   │test                               │"
"   │desc                               │"
"   │┌❌ subtask 2.2───────────────────┐│" Hidden by multi-width symbols: [(6, " ")]
"   ││📅 2016-07-08 09:10:11           ││" Hidden by multi-width symbols: [(6, " ")]
"   ││#tag #tag2                       ││"
"   ││test                             ││"
"   ││desc                             ││"
"   │└─────────────────────────────────┘│"
"   └───────────────────────────────────┘"
"                                        "
"                                        "
"                                        "
"                                        "
